README.md
# Tencent Serverless Http

[![Build Status](https://travis-ci.com/serverless-plus/tencent-serverless-http.svg?branch=master)](https://travis-ci.com/serverless-plus/tencent-serverless-http)
[![npm](https://img.shields.io/npm/v/tencent-serverless-http.svg)](http://www.npmtrends.com/tencent-serverless-http)
[![NPM downloads](https://img.shields.io/npm/dm/tencent-serverless-http.svg)](http://www.npmtrends.com/tencent-serverless-http)

This project is a fork of
[aws-serverless-express](https://github.com/awslabs/aws-serverless-express.git),
and modify for tencent cloud scf.

## Install

```bash
npm install tencent-serverless-http
```

## Usage

```js
// handler.js
'use strict';
const tencentServerlessHttp = require('tencent-serverless-http');
const app = require('./app');
const server = tencentServerlessHttp.createServer(app);

exports.handler = (event, context) => {
  tencentServerlessHttp.proxy(server, event, context);
};
```

## License

MIT License

Copyright (c) 2020 Serverless Plus


CHANGELOG.md
## [1.3.2](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.3.1...v1.3.2) (2021-05-20)


### Bug Fixes

* set isBase64Encoded when content-encoding ([d3d23a4](https://github.com/serverless-plus/tencent-serverless-http/commit/d3d23a4))

## [1.3.1](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.3.0...v1.3.1) (2020-09-17)


### Bug Fixes

* remove protocol request option ([935fd3d](https://github.com/serverless-plus/tencent-serverless-http/commit/935fd3d))

# [1.3.0](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.2.0...v1.3.0) (2020-09-16)


### Features

* remove middleware ([e40dbea](https://github.com/serverless-plus/tencent-serverless-http/commit/e40dbea))

# [1.2.0](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.1.0...v1.2.0) (2020-05-13)


### Features

* remove event and context header ([8b03cbf](https://github.com/serverless-plus/tencent-serverless-http/commit/8b03cbf))

# [1.1.0](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.0.7...v1.1.0) (2020-04-24)


### Features

* remove unuse remarks ([7ee5e67](https://github.com/serverless-plus/tencent-serverless-http/commit/7ee5e67))

## [1.0.7](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.0.6...v1.0.7) (2020-03-17)


### Bug Fixes

* remove module name prefix ([2222497](https://github.com/serverless-plus/tencent-serverless-http/commit/22224979dc4f40a775cc724aa0111013f61c5a7a))

## [1.0.5](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.0.4...v1.0.5) (2020-02-09)


### Bug Fixes

* get tencent api query ([6e740ab](https://github.com/serverless-plus/tencent-serverless-http/commit/6e740ab))

## [1.0.4](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.0.3...v1.0.4) (2020-01-16)


### Bug Fixes

* update lowest node version to 8 ([654b6b8](https://github.com/serverless-plus/tencent-serverless-http/commit/654b6b8))

## [1.0.1](https://github.com/serverless-plus/tencent-serverless-http/compare/v1.0.0...v1.0.1) (2020-01-15)


### Bug Fixes

* update readme ([bc3491e](https://github.com/serverless-plus/tencent-serverless-http/commit/bc3491e))


__tests__/unit.js
'use strict'

const path = require('path')

const awsServerlessExpress = require('../index')

test('getPathWithQueryStringParams: no params', () => {
  const event = {
    path: '/foo/bar'
  }
  const pathWithQueryStringParams = awsServerlessExpress.getPathWithQueryStringParams(event)
  expect(pathWithQueryStringParams).toEqual('/foo/bar')
})

test('getPathWithQueryStringParams: 1 param', () => {
  const event = {
    path: '/foo/bar',
    queryString: {
      'bizz': 'bazz'
    }
  }
  const pathWithQueryStringParams = awsServerlessExpress.getPathWithQueryStringParams(event)
  expect(pathWithQueryStringParams).toEqual('/foo/bar?bizz=bazz')
})

test('getPathWithQueryStringParams: to be url-encoded param', () => {
  const event = {
    path: '/foo/bar',
    queryString: {
      'redirect_uri': 'http://lvh.me:3000/cb'
    }
  }
  const pathWithQueryStringParams = awsServerlessExpress.getPathWithQueryStringParams(event)
  expect(pathWithQueryStringParams).toEqual('/foo/bar?redirect_uri=http%3A%2F%2Flvh.me%3A3000%2Fcb')
})

test('getPathWithQueryStringParams: 2 params', () => {
  const event = {
    path: '/foo/bar',
    queryString: {
      'bizz': 'bazz',
      'buzz': 'bozz'
    }
  }
  const pathWithQueryStringParams = awsServerlessExpress.getPathWithQueryStringParams(event)
  expect(pathWithQueryStringParams).toEqual('/foo/bar?bizz=bazz&buzz=bozz')
})

function mapApiGatewayEventToHttpRequest (headers) {
  const event = {
    path: '/foo',
    httpMethod: 'GET',
    body: 'Hello serverless!',
    headers
  }
  const eventClone = JSON.parse(JSON.stringify(event))
  delete eventClone.body
  const context = {
    'foo': 'bar'
  }
  const socketPath = '/tmp/server0.sock'
  const httpRequest = awsServerlessExpress.mapApiGatewayEventToHttpRequest(event, context, socketPath)

  return {httpRequest, eventClone, context}
}

test('mapApiGatewayEventToHttpRequest: with headers', () => {
  const r = mapApiGatewayEventToHttpRequest({'x-foo': 'foo'})

  expect(r.httpRequest).toEqual({
    method: 'GET',
    path: '/foo',
    headers: {
      'x-foo': 'foo',
      'Content-Length': Buffer.byteLength('Hello serverless!')
    },
    socketPath: '/tmp/server0.sock'
  })
})

test('mapApiGatewayEventToHttpRequest: without headers', () => {
  const r = mapApiGatewayEventToHttpRequest()

  expect(r.httpRequest).toEqual({
    method: 'GET',
    path: '/foo',
    headers: {
      'Content-Length': Buffer.byteLength('Hello serverless!')
    },
    socketPath: '/tmp/server0.sock'
  })
})

test('getSocketPath', () => {
  const socketPath = awsServerlessExpress.getSocketPath('12345abcdef')
  const isWin = process.platform === 'win32'
  const expectedSocketPath = isWin ? path.join('\\\\?\\\\pipe\\\\', process.cwd(), 'server-12345abcdef') : '/tmp/server-12345abcdef.sock'
  expect(socketPath).toBe(expectedSocketPath)
})

const PassThrough = require('stream').PassThrough

class MockResponse extends PassThrough {
  constructor (statusCode, headers, body) {
    super()
    this.statusCode = statusCode
    this.headers = headers || {}
    this.write(body)
    this.end()
  }
}

class MockServer {
  constructor (binaryTypes) {
    this._binaryTypes = binaryTypes || []
  }
}

class MockContext {
  constructor (resolve) {
    this.resolve = resolve
  }
  succeed (successResponse) {
    this.resolve(successResponse)
  }
}

describe('forwardConnectionErrorResponseToApiGateway', () => {
  test('responds with 502 status', () => {
    return new Promise(
      (resolve, reject) => {
        const context = new MockContext(resolve)
        const contextResolver = {
          succeed: (p) => context.succeed(p.response)
        }
        awsServerlessExpress.forwardConnectionErrorResponseToApiGateway('ERROR', contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 502,
      body: '',
      headers: {}
    }))
  })
})

describe('forwardLibraryErrorResponseToApiGateway', () => {
  test('responds with 500 status', () => {
    return new Promise(
      (resolve, reject) => {
        const context = new MockContext(resolve)
        const contextResolver = {
          succeed: (p) => context.succeed(p.response)
        }
        awsServerlessExpress.forwardLibraryErrorResponseToApiGateway('ERROR', contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 500,
      body: '',
      headers: {}
    }))
  })
})

function getContextResolver (resolve) {
  const context = new MockContext(resolve)
  const contextResolver = {
    succeed: (p) => context.succeed(p.response)
  }

  return contextResolver
}
describe('forwardResponseToApiGateway: header handling', () => {
  test('multiple headers with the same name get transformed', () => {
    const server = new MockServer()
    const headers = {'foo': ['bar', 'baz'], 'Set-Cookie': ['bar', 'baz']}
    const body = 'hello world'
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: body,
      headers: { foo: 'bar,baz', 'Set-Cookie': ['bar', 'baz'] },
      isBase64Encoded: false
    }))
  })
})

describe('forwardResponseToApiGateway: content-type encoding', () => {
  test('content-type header missing', () => {
    const server = new MockServer()
    const headers = {'foo': 'bar'}
    const body = 'hello world'
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: body,
      headers: headers,
      isBase64Encoded: false
    }))
  })

  test('content-type image/jpeg base64 encoded', () => {
    const server = new MockServer(['image/jpeg'])
    const headers = {'content-type': 'image/jpeg'}
    const body = 'hello world'
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: Buffer.from(body).toString('base64'),
      headers: headers,
      isBase64Encoded: true
    }))
  })

  test('content-type application/json', () => {
    const server = new MockServer()
    const headers = {'content-type': 'application/json'}
    const body = JSON.stringify({'hello': 'world'})
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: body,
      headers: headers,
      isBase64Encoded: false
    }))
  })

  test('wildcards in binary types array', () => {
    const server = new MockServer(['image/*'])
    const headers = {'content-type': 'image/jpeg'}
    const body = 'hello world'
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: Buffer.from(body).toString('base64'),
      headers: headers,
      isBase64Encoded: true
    }))
  })

  test('extensions in binary types array', () => {
    const server = new MockServer(['.png'])
    const headers = {'content-type': 'image/png'}
    const body = 'hello world'
    const response = new MockResponse(200, headers, body)
    return new Promise(
      (resolve, reject) => {
        const contextResolver = getContextResolver(resolve)
        awsServerlessExpress.forwardResponseToApiGateway(server, response, contextResolver)
      }
    ).then(successResponse => expect(successResponse).toEqual({
      statusCode: 200,
      body: Buffer.from(body).toString('base64'),
      headers: headers,
      isBase64Encoded: true
    }))
  })
})

describe('makeResolver', () => {
  test('CONTEXT_SUCCEED (specified)', () => {
    return new Promise(
      (resolve, reject) => {
        const context = new MockContext(resolve)
        const contextResolver = awsServerlessExpress.makeResolver({
          context,
          resolutionMode: 'CONTEXT_SUCCEED'
        })

        return contextResolver.succeed({
          response: 'success'
        })
      }).then(successResponse => expect(successResponse).toEqual('success'))
  })

  test('CALLBACK', () => {
    const callback = (e, response) => response
    const callbackResolver = awsServerlessExpress.makeResolver({
      callback,
      resolutionMode: 'CALLBACK'
    })
    const successResponse = callbackResolver.succeed({
      response: 'success'
    })

    expect(successResponse).toEqual('success')
  })

  test('PROMISE', () => {
    return new Promise((resolve, reject) => {
      const promise = {
        resolve,
        reject
      }
      const promiseResolver = awsServerlessExpress.makeResolver({
        promise,
        resolutionMode: 'PROMISE'
      })

      return promiseResolver.succeed({
        response: 'success'
      })
    }).then(successResponse => {
      expect(successResponse).toEqual('success')
    })
  })
})


__tests__/middleware.js
'use strict'
const tencentServerlessNodejsMiddleware = require('../middleware')
const eventContextMiddleware = tencentServerlessNodejsMiddleware.eventContext
const mockNext = () => true
const generateMockReq = () => {
  return {
    headers: {
      'x-apigateway-event': encodeURIComponent(JSON.stringify({
        path: '/foo/bar',
        queryStringParameters: {
          foo: '🖖',
          bar: '~!@#$%^&*()_+`-=;\':",./<>?`'
        }
      })),
      'x-apigateway-context': encodeURIComponent(JSON.stringify({foo: 'bar'}))
    }
  }
}
const mockRes = {}

test('defaults', () => {
  const req = generateMockReq()
  const originalHeaders = Object.assign({}, req.headers)

  eventContextMiddleware()(req, mockRes, mockNext)

  expect(req.apiGateway.event).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-event'])))
  expect(req.apiGateway.context).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-context'])))
  expect(req.headers['x-apigateway-event']).toBe(undefined)
  expect(req.headers['x-apigateway-context']).toBe(undefined)
})

test('options.reqPropKey', () => {
  const req = generateMockReq()
  const originalHeaders = Object.assign({}, req.headers)

  eventContextMiddleware({ reqPropKey: '_apiGateway' })(req, mockRes, mockNext)

  expect(req._apiGateway.event).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-event'])))
  expect(req._apiGateway.context).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-context'])))
  expect(req.headers['x-apigateway-event']).toBe(undefined)
  expect(req.headers['x-apigateway-context']).toBe(undefined)
})

test('options.deleteHeaders = false', () => {
  const req = generateMockReq()
  const originalHeaders = Object.assign({}, req.headers)

  eventContextMiddleware({ deleteHeaders: false })(req, mockRes, mockNext)

  expect(req.apiGateway.event).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-event'])))
  expect(req.apiGateway.context).toEqual(JSON.parse(decodeURIComponent(originalHeaders['x-apigateway-context'])))
  expect(req.headers['x-apigateway-event']).toEqual(originalHeaders['x-apigateway-event'])
  expect(req.headers['x-apigateway-context']).toEqual(originalHeaders['x-apigateway-context'])
})

test('Missing x-apigateway-event', () => {
  const req = generateMockReq()
  delete req.headers['x-apigateway-event']

  eventContextMiddleware({ deleteHeaders: false })(req, mockRes, mockNext)

  expect(req.apiGateway).toBe(undefined)
})

test('Missing x-apigateway-context', () => {
  const req = generateMockReq()
  delete req.headers['x-apigateway-context']

  eventContextMiddleware({ deleteHeaders: false })(req, mockRes, mockNext)

  expect(req.apiGateway).toBe(undefined)
})


src/report.js
const { Capi } = require('@tencent-sdk/capi')

function Report (event = {}, context = {}, debugOptions) {
  this.start = new Date()
  const { httpMethod: method, path } = event
  if (!method || !path) {
    this.canReport = false
    console.warn('No method or path in event.')
    return
  }
  this.method = method.toUpperCase()
  this.path = str2hex(path)
  const ServiceType = 'monitor'
  const {
    Region = 'ap-guangzhou',
    function_name: FunctionName,
    function_version: Version = '$latest',
    namespace: Namespace = 'default'
  } = context
  const environment = JSON.parse(context.environment || '{}')
  const {
    TENCENTCLOUD_SECRETID: SecretId,
    TENCENTCLOUD_SECRETKEY: SecretKey,
    TENCENTCLOUD_SESSIONTOKEN: Token
  } = environment

  if (!SecretId || !SecretKey) {
    // Don't report data if no SecretId or SecretKey
    this.canReport = false
    console.warn('No SecretId or SecretKey in environment parameters.')
    return
  }

  this.canReport = true
  this.client = new Capi({
    Region,
    SecretId,
    SecretKey,
    Token,
    ServiceType
  })
  this.commonParams = {
    Version: '2018-07-24',
    AnnounceInstance: `${Namespace}|${FunctionName}|${Version}`
  }
  this.debugOptions = debugOptions || {
    debug: false,
    host: 'monitor.tencentcloudapi.com'
  }
}

Report.prototype.reportRequest = function () {
  if (!this.canReport) {
    return false
  }
  const Metrics = [
    { MetricName: 'request', Value: 1 },
    { MetricName: `${this.method}_${this.path}`, Value: 1 }
  ]
  const _this = this
  this.client.request(
    {
      Action: 'PutMonitorData',
      Metrics,
      ...this.commonParams
    },
    this.debugOptions,
    true
  ).then(() => {
    // console.log(data)
  }, () => {
    _this.canReport = false
  })
}

Report.prototype.reportResponse = async function (response) {
  if (!this.canReport) {
    return false
  }
  const latency = new Date() - this.start
  const keyPrefix = `${this.method}_${this.path}`
  let { statusCode } = response.response || {}
  statusCode = statusCode ? String(statusCode) : ''
  const Metrics = [
    { MetricName: 'latency', Value: latency },
    { MetricName: keyPrefix + '_latency', Value: latency },
    { MetricName: keyPrefix + '_' + statusCode, Value: 1 }
  ]
  if (statusCode.startsWith('4')) {
    Metrics.push({ MetricName: '4xx', Value: 1 })
    Metrics.push({ MetricName: 'error', Value: 1 })
  } else if (statusCode.startsWith('5')) {
    Metrics.push({ MetricName: '5xx', Value: 1 })
    Metrics.push({ MetricName: 'error', Value: 1 })
  }

  await this.client.request(
    {
      Action: 'PutMonitorData',
      Metrics,
      ...this.commonParams
    },
    this.debugOptions,
    true
  )
}

// 字符串转16进制
function str2hex (str) {
  if (str === '') {
    return ''
  }
  let arr = []
  for (let i = 0; i < str.length; i++) {
    arr.push(str.charCodeAt(i).toString(16))
  }
  return arr.join('')
}

module.exports = exports = Report


examples/basic-starter/app.js
'use strict'
const path = require('path')
const express = require('express')
const bodyParser = require('body-parser')
const cors = require('cors')
const compression = require('compression')
const app = express()
const router = express.Router()

app.set('view engine', 'pug')

if (process.env.NODE_ENV === 'test') {
  // NOTE: tencent-serverless-http uses this app for its integration tests
  // and only applies compression to the /sam endpoint during testing.
  router.use('/sam', compression())
} else {
  router.use(compression())
}

router.use(cors())
router.use(bodyParser.json())
router.use(bodyParser.urlencoded({ extended: true }))

// NOTE: tests can't find the views directory without this
app.set('views', path.join(__dirname, 'views'))

router.get('/', (req, res) => {
  res.render('index', {
    apiUrl: req.apiGateway
      ? `https://${req.apiGateway.event.headers.Host}/${req.apiGateway.event.requestContext.stage}`
      : 'http://localhost:3000'
  })
})

router.get('/sam', (req, res) => {
  res.sendFile(`${__dirname}/sam-logo.png`)
})

router.get('/users', (req, res) => {
  res.json(users)
})

router.get('/users/:userId', (req, res) => {
  const user = getUser(req.params.userId)

  if (!user) return res.status(404).json({})

  return res.json(user)
})

router.post('/users', (req, res) => {
  const user = {
    id: ++userIdCounter,
    name: req.body.name
  }
  users.push(user)
  res.status(201).json(user)
})

router.put('/users/:userId', (req, res) => {
  const user = getUser(req.params.userId)

  if (!user) return res.status(404).json({})

  user.name = req.body.name
  res.json(user)
})

router.delete('/users/:userId', (req, res) => {
  const userIndex = getUserIndex(req.params.userId)

  if (userIndex === -1) return res.status(404).json({})

  users.splice(userIndex, 1)
  res.json(users)
})

const getUser = (userId) => users.find((u) => u.id === parseInt(userId))
const getUserIndex = (userId) =>
  users.findIndex((u) => u.id === parseInt(userId))

// Ephemeral in-memory data store
const users = [
  {
    id: 1,
    name: 'Joe'
  },
  {
    id: 2,
    name: 'Jane'
  }
]
let userIdCounter = users.length

// The tencent-serverless-http library creates a server and listens on a Unix
// Domain Socket for you, so you can remove the usual call to app.listen.
// app.listen(3000)
app.use('/', router)

// Export your express server so you can import it in the lambda function.
module.exports = app


examples/basic-starter/lambda.js
'use strict'
const awsServerlessExpress = require(process.env.NODE_ENV === 'test' ? '../../index' : 'tencent-serverless-http')
const app = require('./app')

// NOTE: If you get ERR_CONTENT_DECODING_FAILED in your browser, this is likely
// due to a compressed response (e.g. gzip) which has not been handled correctly
// by tencent-serverless-http and/or API Gateway. Add the necessary MIME types to
// binaryMimeTypes below, then redeploy (`npm run package-deploy`)
const binaryMimeTypes = [
  'application/javascript',
  'application/json',
  'application/octet-stream',
  'application/xml',
  'font/eot',
  'font/opentype',
  'font/otf',
  'image/jpeg',
  'image/png',
  'image/svg+xml',
  'text/comma-separated-values',
  'text/css',
  'text/html',
  'text/javascript',
  'text/plain',
  'text/text',
  'text/xml'
]
const server = awsServerlessExpress.createServer(app, null, binaryMimeTypes)

exports.handler = (event, context) => awsServerlessExpress.proxy(server, event, context)


examples/basic-starter/handler.js
'use strict'
const serverlessExpress = require(process.env.NODE_ENV === 'test' ? '../../index' : 'tencent-serverless-http')
const app = require('./app')

// NOTE: If you get ERR_CONTENT_DECODING_FAILED in your browser, this is likely
// due to a compressed response (e.g. gzip) which has not been handled correctly
// by tencent-serverless-http and/or API Gateway. Add the necessary MIME types to
// binaryMimeTypes below, then redeploy (`npm run package-deploy`)
const binaryMimeTypes = [
  'application/javascript',
  'application/json',
  'application/octet-stream',
  'application/xml',
  'font/eot',
  'font/opentype',
  'font/otf',
  'image/jpeg',
  'image/png',
  'image/svg+xml',
  'text/comma-separated-values',
  'text/css',
  'text/html',
  'text/javascript',
  'text/plain',
  'text/text',
  'text/xml'
]
const server = serverlessExpress.createServer(app, null, binaryMimeTypes)

exports.handler = (event, context) => serverlessExpress.proxy(server, event, context)


src/index.js
'use strict'
const http = require('http')
const url = require('url')
const isType = require('type-is')

function getPathWithQueryStringParams (event) {
  return url.format({ pathname: event.path, query: event.queryString })
}
function getEventBody (event) {
  return Buffer.from(event.body, event.isBase64Encoded ? 'base64' : 'utf8')
}

function clone (json) {
  return JSON.parse(JSON.stringify(json))
}

function getContentType (params) {
  // only compare mime type; ignore encoding part
  return params.contentTypeHeader ? params.contentTypeHeader.split(';')[0] : ''
}

function isContentTypeBinaryMimeType (params) {
  return params.binaryMimeTypes.length > 0 && !!isType.is(params.contentType, params.binaryMimeTypes)
}

function mapApiGatewayEventToHttpRequest (event, context, socketPath) {
  const headers = Object.assign({}, event.headers)

  // NOTE: API Gateway is not setting Content-Length header on requests even when they have a body
  if (event.body && !headers['Content-Length']) {
    const body = getEventBody(event)
    headers['Content-Length'] = Buffer.byteLength(body)
  }

  const clonedEventWithoutBody = clone(event)
  delete clonedEventWithoutBody.body

  return {
    method: event.httpMethod,
    path: getPathWithQueryStringParams(event),
    headers,
    socketPath
  }
}

function forwardResponseToApiGateway (server, response, resolver) {
  let buf = []

  response
    .on('data', (chunk) => buf.push(chunk))
    .on('end', () => {
      const bodyBuffer = Buffer.concat(buf)
      const statusCode = response.statusCode
      const headers = response.headers

      // chunked transfer not currently supported by API Gateway
      /* istanbul ignore else */
      if (headers['transfer-encoding'] === 'chunked') {
        delete headers['transfer-encoding']
      }

      Object.keys(headers)
        .forEach(h => {
          if (Array.isArray(headers[h])) {
            if (h.toLowerCase() !== 'set-cookie') {
              headers[h] = headers[h].join(',')
            }
          }
        })

      const contentType = getContentType({ contentTypeHeader: headers['content-type'] })
      let isBase64Encoded = isContentTypeBinaryMimeType({ contentType, binaryMimeTypes: server._binaryTypes })
      if (['gzip', 'compress', 'deflate', 'br'].indexOf(headers['content-encoding']) !== -1) {
        isBase64Encoded = true
      }
      const body = bodyBuffer.toString(isBase64Encoded ? 'base64' : 'utf8')
      const successResponse = {statusCode, body, headers, isBase64Encoded}

      resolver.succeed({ response: successResponse })
    })
}

function forwardConnectionErrorResponseToApiGateway (error, resolver) {
  console.log('ERROR: tencent-serverless-http connection error')
  console.error(error)
  const errorResponse = {
    statusCode: 502, // "DNS resolution, TCP level errors, or actual HTTP parse errors" - https://nodejs.org/api/http.html#http_http_request_options_callback
    body: '',
    headers: {}
  }

  resolver.succeed({ response: errorResponse })
}

function forwardLibraryErrorResponseToApiGateway (error, resolver) {
  console.log('ERROR: tencent-serverless-http error')
  console.error(error)
  const errorResponse = {
    statusCode: 500,
    body: '',
    headers: {}
  }

  resolver.succeed({ response: errorResponse })
}

function forwardRequestToNodeServer (server, event, context, resolver) {
  try {
    const requestOptions = mapApiGatewayEventToHttpRequest(event, context, getSocketPath(server._socketPathSuffix))
    const req = http.request(requestOptions, (response) => forwardResponseToApiGateway(server, response, resolver))
    if (event.body) {
      const body = getEventBody(event)

      req.write(body)
    }

    req.on('error', (error) => forwardConnectionErrorResponseToApiGateway(error, resolver))
      .end()
  } catch (error) {
    forwardLibraryErrorResponseToApiGateway(error, resolver)
    return server
  }
}

function startServer (server) {
  return server.listen(getSocketPath(server._socketPathSuffix))
}

function getSocketPath (socketPathSuffix) {
  /* istanbul ignore if */ /* only running tests on Linux; Window support is for local dev only */
  if (/^win/.test(process.platform)) {
    const path = require('path')
    return path.join('\\\\?\\pipe', process.cwd(), `server-${socketPathSuffix}`)
  } else {
    return `/tmp/server-${socketPathSuffix}.sock`
  }
}

function getRandomString () {
  return Math.random().toString(36).substring(2, 15)
}

function createServer (requestListener, serverListenCallback, binaryTypes) {
  const server = http.createServer(requestListener)

  server._socketPathSuffix = getRandomString()
  server._binaryTypes = binaryTypes ? binaryTypes.slice() : []
  server.on('listening', () => {
    server._isListening = true

    if (serverListenCallback) serverListenCallback()
  })
  server.on('close', () => {
    server._isListening = false
  })
    .on('error', (error) => {
      /* istanbul ignore else */
      if (error.code === 'EADDRINUSE') {
        console.warn(`WARNING: Attempting to listen on socket ${getSocketPath(server._socketPathSuffix)}, but it is already in use. This is likely as a result of a previous invocation error or timeout. Check the logs for the invocation(s) immediately prior to this for root cause, and consider increasing the timeout and/or cpu/memory allocation if this is purely as a result of a timeout. tencent-serverless-http will restart the Node.js server listening on a new port and continue with this request.`)
        server._socketPathSuffix = getRandomString()
        return server.close(() => startServer(server))
      } else {
        console.log('ERROR: server error')
        console.error(error)
      }
    })

  return server
}

function proxy (server, event, context, resolutionMode, callback) {
  // DEPRECATED: Legacy support
  if (!resolutionMode) {
    const resolver = makeResolver({ context, resolutionMode: 'CONTEXT_SUCCEED' })
    if (server._isListening) {
      forwardRequestToNodeServer(server, event, context, resolver)
      return server
    } else {
      return startServer(server)
        .on('listening', () => proxy(server, event, context))
    }
  }

  return {
    promise: new Promise((resolve, reject) => {
      const promise = {
        resolve,
        reject
      }
      const resolver = makeResolver({
        context,
        callback,
        promise,
        resolutionMode
      })

      if (server._isListening) {
        forwardRequestToNodeServer(server, event, context, resolver)
      } else {
        startServer(server)
          .on('listening', () => forwardRequestToNodeServer(server, event, context, resolver))
      }
    })
  }
}

function makeResolver (params/* {
  context,
  callback,
  promise,
  resolutionMode
} */) {
  return {
    succeed: (params2/* {
      response
    } */) => {
      if (params.resolutionMode === 'CONTEXT_SUCCEED') return params.context.succeed(params2.response)
      if (params.resolutionMode === 'CALLBACK') return params.callback(null, params2.response)
      if (params.resolutionMode === 'PROMISE') return params.promise.resolve(params2.response)
    }
  }
}

exports.createServer = createServer
exports.proxy = proxy

/* istanbul ignore else */
if (process.env.NODE_ENV === 'test') {
  exports.getPathWithQueryStringParams = getPathWithQueryStringParams
  exports.mapApiGatewayEventToHttpRequest = mapApiGatewayEventToHttpRequest
  exports.forwardResponseToApiGateway = forwardResponseToApiGateway
  exports.forwardConnectionErrorResponseToApiGateway = forwardConnectionErrorResponseToApiGateway
  exports.forwardLibraryErrorResponseToApiGateway = forwardLibraryErrorResponseToApiGateway
  exports.forwardRequestToNodeServer = forwardRequestToNodeServer
  exports.startServer = startServer
  exports.getSocketPath = getSocketPath
  exports.makeResolver = makeResolver
}
